aws/hll/aws-s3-transfer-manager/src/download/object_meta.rs

```diff
@@ -0,0 +1,147 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+use aws_sdk_s3::operation::{get_object::GetObjectOutput, head_object::HeadObjectOutput};
+
+// TODO(design): how many of these fields should we expose?
+// TODO(docs): Document fields
+
+/// Object metadata other than the body that can be set from either `GetObject` or `HeadObject`
+#[derive(Debug, Clone)]
+pub struct ObjectMetadata {
+    pub delete_marker: Option<bool>,
+    pub accept_ranges: Option<String>,
+    pub expiration: Option<String>,
+    pub restore: Option<String>,
+    pub last_modified: Option<::aws_smithy_types::DateTime>,
+    pub content_length: Option<i64>,
+    pub e_tag: Option<String>,
+    pub checksum_crc32: Option<String>,
+    pub checksum_crc32_c: Option<String>,
+    pub checksum_sha1: Option<String>,
+    pub checksum_sha256: Option<String>,
+    pub missing_meta: Option<i32>,
+    pub version_id: Option<String>,
+    pub cache_control: Option<String>,
+    pub content_disposition: Option<String>,
+    pub content_encoding: Option<String>,
+    pub content_language: Option<String>,
+    pub content_range: Option<String>,
+    pub content_type: Option<String>,
+    pub expires: Option<::aws_smithy_types::DateTime>,
+    pub website_redirect_location: Option<String>,
+    pub server_side_encryption: Option<aws_sdk_s3::types::ServerSideEncryption>,
+    pub metadata: Option<::std::collections::HashMap<String, String>>,
+    pub sse_customer_algorithm: Option<String>,
+    pub sse_customer_key_md5: Option<String>,
+    pub ssekms_key_id: Option<String>,
+    pub bucket_key_enabled: Option<bool>,
+    pub storage_class: Option<aws_sdk_s3::types::StorageClass>,
+    pub request_charged: Option<aws_sdk_s3::types::RequestCharged>,
+    pub replication_status: Option<aws_sdk_s3::types::ReplicationStatus>,
+    pub parts_count: Option<i32>,
+    pub tag_count: Option<i32>,
+    pub object_lock_mode: Option<aws_sdk_s3::types::ObjectLockMode>,
+    pub object_lock_retain_until_date: Option<::aws_smithy_types::DateTime>,
+    pub object_lock_legal_hold_status: Option<aws_sdk_s3::types::ObjectLockLegalHoldStatus>,
+}
+
+impl ObjectMetadata {
+    /// The total object size
+    pub(crate) fn total_size(&self) -> u64 {
+        match (self.content_length, self.content_range.as_ref()) {
+            (_, Some(range)) => {
+                let total = range.split_once('/').map(|x| x.1).expect("content range total");
+                total.parse().expect("valid range total")
+            }
+            (Some(length), None) => length as u64,
+            (None, None) => panic!("total object size cannot be calculated without either content length or content range headers")
+        }
+    }
+}
+
+impl From<GetObjectOutput> for ObjectMetadata {
```

Comment:
Before my last project wrapped up I was doing a lot of these `From` impls for types that shared a lot of fields with the same name/type and I really wanted a macro to do most of this for me. If we're going to be doing a lot of these might be a macro worth investing in now.

Reply:
Yeah unsure at the moment exactly what to expose and how often we'll need this (but agree if we find ourselves doing this more we should consider that option).